graph-builder/src/registry.rs

```diff
@@ -37,23 +45,109 @@ impl Into<cincinnati::Release> for Release {
     }
 }
 
+type RegistryClient = dkregistry::v2::Client;
+
+fn registry_trim(src: &str) -> &str {
+    src.trim_left_matches("https://")
+        .trim_left_matches("http://")
+}
+
+fn get_credentials_or_empty_reader(
+    credentials_file: &str,
+) -> Result<Box<dyn std::io::Read>, Error> {
+    if credentials_file != "" {
+        match File::open(credentials_file) {
+            Ok(file) => Ok(Box::new(file)),
+            Err(err) => bail!(err),
+        }
+    } else {
+        Ok(Box::new(std::io::empty()))
+    }
+}
+
+fn get_authenticated_client(
+    tcore: &mut Core,
+    host: &str,
+    image: &str,
+    credentials_file: &str,
+) -> Result<RegistryClient, Error> {
+    let mut registry_client = dkregistry::v2::Client::configure(&tcore.handle())
+        .registry(host)
+        .insecure_registry(false)
+        .read_credentials(get_credentials_or_empty_reader(credentials_file)?)
+        .build()
+        .map_err(|e| format_err!("{}", e))?;
+
+    let fut_check_v2_prelogin = registry_client
+        .is_v2_supported()
+        .map_err(|e| format_err!("{}", e))?;
+    if !tcore
+        .run(fut_check_v2_prelogin)
+        .map_err(|e| format_err!("{}", e))?
+    {
+        return Err(format_err!("API v2 not supported by {}", host));
+    };
+
+    let fut_check_auth_prelogin = registry_client
+        .is_auth(None)
+        .map_err(|e| format_err!("{}", e))?;
+    let logged_in = tcore
+        .run(fut_check_auth_prelogin)
+        .map_err(|e| format_err!("{}", e))?;
+    if logged_in {
+        return Err(format_err!("no login performed, but already authenticated"));
+    }
+
+    let fut_login = registry_client
+        .login(vec![&format!("repository:{}:pull", image)])
```

Comment:
Does this method require a Vec? I'd avoid heap allocations if possible.

Reply:
Nope, I was just sloppy in the original API. Current master already takes a slice.

Reply:
I've got that fixed locally after a clippy run. I'll push in a bit to update this PR